app/coffee/DeleteQueueManager.coffee

```diff
@@ -0,0 +1,62 @@
+RedisManager = require "./RedisManager"
+ProjectManager = require "./ProjectManager"
+logger = require "logger-sharelatex"
+metrics = require "./Metrics"
+async = require "async"
+
+# Maintain a sorted set of project flushAndDelete requests, ordered by timestamp
+# (ZADD), and process them from oldest to newest. A flushAndDelete request comes
+# from real-time and is triggered when a user leaves a project. 
+#
+# The aim is to remove the project from redis 5 minutes after the last request
+# if there has been no activity (document updates) in that time.  If there is
+# activity we can expect a further flushAndDelete request when the editing user
+# leaves the project. 
+#
+# If a new flushAndDelete request comes in while an existing request is already
+# in the queue we update the timestamp as we can postpone flushing further.
+#
+# Documents are processed by checking the queue, seeing if the first entry is
+# older than 5 minutes, and popping it from the queue in that case.
+
+module.exports = DeleteQueueManager =
+    flushAndDeleteOldProjects: (options, callback) ->
+        startTime = Date.now()
+        count = 0
+
+        flushProjectIfNotModified = (project_id, flushTimestamp, cb) ->
+            ProjectManager.getProjectDocsTimestamps project_id, (err, timestamps) ->
+                return callback(err) if err?
+                if !timestamps?
+                    logger.log {project_id}, "skipping flush of queued project - no timestamps"
+                    return cb()
+                # are any of the timestamps newer than the time the project was flushed?
+                for timestamp in timestamps or [] when timestamp > flushTimestamp
+                    metrics.inc "queued-delete-skipped"
+                    logger.debug {project_id, timestamps, flushTimestamp}, "found newer timestamp, will skip delete"
+                    return cb()
+                logger.log {project_id, flushTimestamp}, "flushing queued project"
+                ProjectManager.flushAndDeleteProjectWithLocks project_id, {skip_history_flush: true}, (err) ->
```

Comment:
Previously, I think we'd only have skipped the history flush if the request was with `?shutdown`, but now I think we'll apply the same to `?background`. This is probably OK. It may mean that more history will build up to be flushed if the user does need it (e.g. if they open history)?

Reply:
That's correct,  we still had project history problems with an unexpected mass disconnection (but not from a graceful real-time pod eviction).  I think the periodic flushing to history is enough and if there is a problem we should look at making it scale if the number of queues gets too big.